library/psa_crypto_se.h

```diff
@@ -0,0 +1,37 @@
+/*
+ *  PSA crypto support for secure element drivers
+ */
+/*  Copyright (C) 2019, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of Mbed TLS (https://tls.mbed.org)
+ */
+
+#ifndef PSA_CRYPTO_SE_H
+#define PSA_CRYPTO_SE_H
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#include "psa/crypto.h"
+#include "psa/crypto_se_driver.h"
+
+/** The maximum number of registered secure element driver lifetimes. */
+#define PSA_MAX_SE_DRIVERS 4
```

Comment:
Yes, this should be in `config.h`. Let's move it there when we move `
PSA_KEY_SLOT_COUNT`.

I imagine most applications will have 2: one for read-only keys and one for rewritable keys, both with the same method table. 4 doesn't seem like such a bad default to me. Besides, at least for PSA platforms, we'll eventually have a compile-time declaration mechanism that can adjust the size to fit.

Reply:
I dont quite understand. You said that you expect 2 to be the common case, but still think 4 is a better choice? If we know that 2 is probably very common, why do we need to take up more space and put a 4? It would be nice to be able to configure this at compile-time, but I also think having a good default is important as many users will probably just skip over this.

Reply:
Let's debate the default later. Configuration is out of scope of this PR.

Reply:
Happy to leave it for later, but I do not think this is out of the scope of the PR since we are just talking about what the default value should be. My point is: you mentioned that the common case is 2 but set the default to 4.

Reply:
Also, +1 to @Patater's comment about the config.h

Reply:
@gilles-peskine-arm , I think that driver registration and enumeration should be  related to HW only, without any dependency on key read/write attributes. If you have 1 SE you have 1 driver to register and to use. Do we have an advantage of registration the same SE twice (1 only for read and 2 for read/write) ?

Reply:
@shelib01 The point is that the application is the entity that knows when the key must be erased. The lifetime of the key combines the physical location (e.g. internal storage, secure element) and the logical location inside that physical location (e.g. directory — but a factory reset subsystem could also enumerate all keys and erase them by their lifetime). Combining the two makes it very cheap to implement. Separating them would make the interface more complex (the driver would still need to expose logical location information) and costlier to implement (more code overhead).

Reply:
If pre-provisioned slots are locked nobody will not be able to erase them anyway, if they are unlocked I can erase it if I want. Why do I need specific status for it ? I have also my own (not -pre- provisioned keys) that should not be deleted in factory reset flow, so it will be confusing . Additional question how you will deal with situation of locking slot in run time . You will change lifetime of this key after slot locking?

Reply:
May this read-only tag can be a pat of key attribute or policy ?

Reply:
If you inject your own keys that shouldn't be deleted on a factory reset, you give them the lifetime `IN_SECURE_ELEMENT_NOT_DELETED_ON_FACTORY_RESET`. It doesn't matter whether the slots are locked.

“Read-only” as a key attribute is not precise enough. Read only to whom? Until what point?

Reply:
You are right read-only name is not good enough. It should be kind of write-once. 
I will be able to  give the same NOT_DELETED_ON_FACTORY also to regular persistent keys ?  
It will be possible to delete these keys not in factory reset ?

Reply:
May be a better name for this lifetime is IN_SECURE_ELEMENT_PRE_PROVISIONED . And deletion will be application decision ?

Reply:
It's a different lifetime `IN_INTERNAL_STORAGE_NOT_DELETED_ON_FACTORY_RESET`. Obviously the factory reset code has to understand these lifetimes. The PSA specification leaves these lifetimes up to each implementation because they depend a lot on the deployment flow. Mbed Crypto doesn't support alternative lifetimes for internal storage yet, but will at some point in the future.